File: system-test/dotnet/Dsl.Core/Scenario/ExecutionResultBuilder.cs
using Dsl.Core.Shared;
using Common;
using Dsl.Core.UseCase.UseCases.Base;
using Driver.Port.Dtos.Error;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Dsl.Core.Scenario
{
    public class ExecutionResultBuilder<TSuccessResponse, TSuccessVerification>
        where TSuccessVerification : ResponseVerification<TSuccessResponse>
    {
        private readonly MyShopUseCaseResult<TSuccessResponse, TSuccessVerification> _result;
        private string? _orderNumber;
        private string? _couponCode;

        internal ExecutionResultBuilder(UseCaseResult<TSuccessResponse, SystemError, TSuccessVerification, SystemErrorFailureVerification> result)
        {
            // Cast to derived type - the result is always a MyShopUseCaseResult at runtime
            _result = (MyShopUseCaseResult<TSuccessResponse, TSuccessVerification>)result;
        }

        public ExecutionResultBuilder<TSuccessResponse, TSuccessVerification> OrderNumber(string? orderNumber)
        {
            _orderNumber = orderNumber;
            return this;
        }

        public ExecutionResultBuilder<TSuccessResponse, TSuccessVerification> CouponCode(string? couponCode)
        {
            _couponCode = couponCode;
            return this;
        }

        public ExecutionResult<TSuccessResponse, TSuccessVerification> Build()
        {
            return new ExecutionResult<TSuccessResponse, TSuccessVerification>(
                _result,
                _orderNumber,
                _couponCode);
        }
    }
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/scenario/ExecutionResultBuilder.java
package com.mycompany.myshop.testkit.dsl.core.scenario;

import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.shared.ResponseVerification;

public class ExecutionResultBuilder<R, V extends ResponseVerification<R>> {
    private final UseCaseResult<R, V> result;
    private String orderNumber;
    private String couponCode;

    public ExecutionResultBuilder(UseCaseResult<R, V> result) {
        this.result = result;
    }

    public ExecutionResultBuilder<R, V> orderNumber(String orderNumber) {
        this.orderNumber = orderNumber;
        return this;
    }

    public ExecutionResultBuilder<R, V> couponCode(String couponCode) {
        this.couponCode = couponCode;
        return this;
    }

    public ExecutionResult<R, V> build() {
        return new ExecutionResult<>(result, orderNumber, couponCode);
    }
}
